### PR TITLE
Unrevert "[Sema] Setter has incorrect mutating-ness inside class-constrained protocol extension"

### PR DESCRIPTION
**Pull request by theblixguy**
This un-reverts #26695. The Core Team accepts this as a [bug fix](https://forums.swift.org/t/class-constrained-protocol-extension-has-incorrect-mutatingness/27962/11).

Resolves [SR-11298](https://bugs.swift.org/browse/SR-11298).

### CODE REVIEW

File: CHANGELOG.md
Comment (jrose-apple) [inline CHANGELOG.md]: Can you include the example of the code that would break?

Comment (jrose-apple) [reply CHANGELOG.md]: And maybe how to work around that.

Comment (theblixguy) [reply CHANGELOG.md]: Sure, added one now.

File: CHANGELOG.md
Comment (jrose-apple) [inline CHANGELOG.md]: Wait, so the default changes to `nonmutating` but you're still allowed to *write* `mutating`? That's confusing but also a very easy answer. Can you add this as a test case to make sure it works?

Comment (theblixguy) [reply CHANGELOG.md]: There's already a test case (see the last one). Do you mean a test case where I have explicitly written `mutating`?

Comment (theblixguy) [reply CHANGELOG.md]: Actually, I _assumed_ we can add `mutating` to the setter (by looking at the fix-it), but I didn't actually try it. 

The fix-it is actually bogus because adding `mutating` will not make the error go away. What actually needs to be done is marking the protocol requirement as `nonmutating` as well, so we need to either not offer the fix-it or tweak it. I suppose I can do that in a follow-up PR.

Comment (jrose-apple) [reply CHANGELOG.md]: I was thinking the workaround is "assign into a temporary var, even though the change won't be reflected elsewhere". We definitely shouldn't offer the fix-it if it doesn't work. Changing the requirement probably isn't acceptable.

Comment (theblixguy) [reply CHANGELOG.md]: Yeah, I've removed the fix-it now for that case.

File: CHANGELOG.md
Comment (jrose-apple) [inline CHANGELOG.md]: I still think this is the wrong recommendation: the extension is constrained because the protocol is not class-bound itself, and that means the mutating-ness of the existing property might be relevant.

Comment (theblixguy) [reply CHANGELOG.md]: Sure, I suppose we can just drop this and just keep the non-working example. I thought it would make sense to offer at least one possible solution (perhaps this could be reworded), but it's probably okay to not offer one.

Comment (jrose-apple) [reply CHANGELOG.md]: No, I mean, the right answer is

```swift
var anotherProperty1: Int {
  get { return someProperty }
  set {
    var mutableSelf = self
    mutableSelf.someProperty = newValue
    assert(mutableSelf === self, "assigning the property really did reset 'self'")
  }
}
```

…though I'm not sure if we want to explain that to people.

Comment (theblixguy) [reply CHANGELOG.md]: I think it's worth adding a solution.

Comment (theblixguy) [reply CHANGELOG.md]: Added it now.

File: CHANGELOG.md
Comment (jrose-apple) [inline CHANGELOG.md]: Nitpick: "mutable variable" is probably easier for people to understand.

**Review by jrose-apple — COMMENTED** [review]
Thanks!

File: CHANGELOG.md
Comment (theblixguy) [reply CHANGELOG.md]: Fixed!

File: lib/AST/Decl.cpp
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: This is very specific but still feels like it's not getting at the interesting bit. How about just "if the `Self` type (`AD->getDeclContext()->getSelfInterfaceType()`) is class-bound (`isAnyClassReferenceType()`)?"

Comment (theblixguy) [reply lib/AST/Decl.cpp]: Sure, do you mean we should replace all of the checks below with a simple "is Self type class-bound or not"?

Comment (jrose-apple) [reply lib/AST/Decl.cpp]: Yeah. If Self is class-bound, you can't insert `mutating`, at least not today, and that covers all the cases we care about introduced by your other change.

Comment (theblixguy) [reply lib/AST/Decl.cpp]: Alright, done!

File: lib/Sema/TypeCheckDecl.cpp
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: I'm just noticing this now, sorry, but this will do all the work even if the immediate context type has reference semantics and it wasn't important to check the extension. Please use early returns.

File: lib/AST/Decl.cpp
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: In retrospect, should this just be `!doesContextHaveValueSemantics(AD->getDeclContext())` instead of what I said? (And whoops on `getSelfInterfaceType` vs. `getSelfTypeInContext`, yours is correct.)

Comment (theblixguy) [reply lib/AST/Decl.cpp]: Hmm `doesContextHaveValueSemantics` is not accessible here though and it also calls `isReferenceTypeImpl()`, which is what `isAnyClassReferenceType()` and `hasReferenceSemantics()` calls (difference between the two that the former doesn't include functions but the latter does).

I think we don't need the `doesContextHaveValueSemantics` check as `getSelfTypeInContext()` will return the mapped `Self` type of the protocol or extension (it would be a `PrimaryArchetypeType`) and then the `isReferenceTypeImpl()` check would look into it to check if it requires class.

What do you think?

Comment (theblixguy) [reply lib/AST/Decl.cpp]: Maybe we can move `doesContextHaveValueSemantics()` into `DeclContext`.

File: lib/Sema/TypeCheckDecl.cpp
Comment (theblixguy) [reply lib/Sema/TypeCheckDecl.cpp]: I think we should also be using `getSelfTypeInContext()` here as well, right? Then we won’t need the extension check!

File: lib/AST/Decl.cpp
Comment (jrose-apple) [reply lib/AST/Decl.cpp]: I think `getSelfTypeInContext()` is going to be more expensive than checking against the generic signature (because the former would actually have to form the type), but either way is okay, as is moving the predicate into DeclContext. @slavapestov, any opinions?

**Comment by jrose-apple** [comment]
Thanks for sticking with this, by the way. Sorry for only noticing one or two things each time—I keep thinking "oh, this is an unrevert and it was basically ready to go in last time so I don't really need to look at it closely".

File: lib/AST/Decl.cpp
Comment (theblixguy) [reply lib/AST/Decl.cpp]: I've moved `doesContextHaveValueSemantics()` into `DeclContext` and switched to `getSelfTypeInContext()`, which is simpler to use, but I can revert back to the generic signature check if it's less expensive.

**Comment by theblixguy** [comment]
Running tests while I wait for approval:

@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please test source compatibility

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please test source compatibility

File: lib/AST/DeclContext.cpp
Comment (jrose-apple) [inline lib/AST/DeclContext.cpp]: Looks like some tabs snuck in.

Comment (theblixguy) [reply lib/AST/DeclContext.cpp]: Whoops! I’ll fix it after the test finishes.

**Comment by theblixguy** [comment]
Alright, all tests have passed :shipit:

**Comment by theblixguy** [comment]
Is this good to merge now? Just need an approval before I do that (and fix the indentation).

File: lib/AST/DeclContext.cpp
Comment (jrose-apple) [reply lib/AST/DeclContext.cpp]: Still not fixed!

Comment (jrose-apple) [reply lib/AST/DeclContext.cpp]: Though I'm not sure I like the idea of testing it here, rather than asserting it and having callers check.

Comment (theblixguy) [reply lib/AST/DeclContext.cpp]: `getSelfTypeInContext` already asserts that, but seems like this check is called from places where we don’t have a type context and thus end up firing the assert (ex: when building the stdlib). 

The code before was calling `getDeclaredInterfaceType` which would just return an `ErrorType` or null if we couldn’t get the type.

Comment (jrose-apple) [reply lib/AST/DeclContext.cpp]: > `getSelfTypeInContext` already asserts that, but seems like this check is called from places where we don’t have a type context and thus end up firing the assert (ex: when building the stdlib).

Yeah, I remember seeing you add it in here. But asking `hasValueSemantics` doesn't actually make sense for a non-type context, and I don't think we should silently accept it by saying "no".

Comment (theblixguy) [reply lib/AST/DeclContext.cpp]: Alright, I have removed the check.

File: lib/Sema/TypeCheckDecl.cpp
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: `FD->isInstanceMember()` already implies `functionDC->isTypeContext()`, no? (here and below)

Comment (theblixguy) [reply lib/Sema/TypeCheckDecl.cpp]: Hmm I am not sure, it doesn't check for a type context internally (but for a context name and the index of the Self parameter). Let me verify

Comment (jrose-apple) [reply lib/Sema/TypeCheckDecl.cpp]: That's the `isInstanceMember` in Parser. The one in ValueDecl does start off by checking for a type context.

Comment (theblixguy) [reply lib/Sema/TypeCheckDecl.cpp]: Oops, you're right. Xcode took me to the wrong place and I didn't notice. I'll remove the extra check now.

Comment (theblixguy) [reply lib/Sema/TypeCheckDecl.cpp]: Done.

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Thanks again for sticking with it!

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
@swift-ci please test source compatibility

**Pull request merged**
